fw/http.c

```diff
@@ -2677,23 +2704,38 @@ tfw_http_expand_hbh(TfwHttpResp *resp, unsigned short status)
 static int
 tfw_http_set_hdr_connection(TfwHttpMsg *hm, unsigned long conn_flg)
 {
+	int r;
 	BUILD_BUG_ON(BIT_WORD(__TFW_HTTP_MSG_M_CONN) != 0);
 	if (((hm->flags[0] & __TFW_HTTP_MSG_M_CONN) == conn_flg)
 	    && (!TFW_STR_EMPTY(&hm->h_tbl->tbl[TFW_HTTP_HDR_CONNECTION]))
-	    && !test_bit(TFW_HTTP_B_CONN_EXTRA, hm->flags))
+	    && !test_bit(TFW_HTTP_B_CONN_EXTRA, hm->flags)
+	    && !test_bit(TFW_HTTP_B_CONN_UPGRADE, hm->flags))
+	{
 		return 0;
+	}
 
 	switch (conn_flg) {
 	case BIT(TFW_HTTP_B_CONN_CLOSE):
 		return TFW_HTTP_MSG_HDR_XFRM(hm, "Connection", "close",
 					     TFW_HTTP_HDR_CONNECTION, 0);
 	case BIT(TFW_HTTP_B_CONN_KA):
-		return TFW_HTTP_MSG_HDR_XFRM(hm, "Connection", "keep-alive",
-					     TFW_HTTP_HDR_CONNECTION, 0);
+		r = TFW_HTTP_MSG_HDR_XFRM(hm, "Connection", "keep-alive",
+					  TFW_HTTP_HDR_CONNECTION, 0);
+		break;
 	default:
-		return TFW_HTTP_MSG_HDR_DEL(hm, "Connection",
+		r = TFW_HTTP_MSG_HDR_DEL(hm, "Connection",
 					    TFW_HTTP_HDR_CONNECTION);
 	}
+
+	if (r < 0)
+		return r;
+
+	if (test_bit(TFW_HTTP_B_UPGRADE_WEBSOCKET, hm->flags)
+	    && test_bit(TFW_HTTP_B_CONN_UPGRADE, hm->flags))
+		r = TFW_HTTP_MSG_HDR_XFRM(hm, "Connection", "upgrade",
+				      TFW_HTTP_HDR_CONNECTION, 1);
```

Comment:
Headers modification is a heavy operation, probably involving memory allocations and skb frarmentation (see https://github.com/tempesta-tech/tempesta/issues/1103), so it'd be more beneficial to do bit more complex condition and call do only one transformation call.

Reply:
That make sense. I'll fix it.